<a href="https://colab.research.google.com/github/therealsamrajya/6th-Sem/blob/main/movie_recommendation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

difflib is used to compare the movie names with the list and find the closest matches. even the spelling is mistake it helps to find the closest ones

Tfidfvectorizer is used to convert textual data to numerical data(feature vectors)

cosine similarity : similarity score is given to the movie which comparing and on that score movies will be recommended

Importing of libraries

In [7]:
import numpy as np
import pandas as pd
import difflib
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

Loading and preprocessing

In [8]:
movie_data = pd.read_csv("/content/movies.csv")

In [9]:
movie_data.head()

index     budget                                    genres  \
0      0  237000000  Action Adventure Fantasy Science Fiction   
1      1  300000000                  Adventure Fantasy Action   
2      2  245000000                    Action Adventure Crime   
3      3  250000000               Action Crime Drama Thriller   
4      4  260000000          Action Adventure Science Fiction   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  culture clash future space war space colony so...                en   
1  ocean drug abuse exotic island east india trad...                en   
2         spy based on novel secret agent sequel mi6                en   
3  dc comics crime fighter terrorist secret ident...                en   
4  based on novel mars medallion space travel pri...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  ... runtime  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577  ...   162.0   
1  Captain Barbossa, long believed to be dead, ha...  139.082615  ...   169.0   
2  A cryptic message from Bond’s past sends him o...  107.376788  ...   148.0   
3  Following the death of District Attorney Harve...  112.312950  ...   165.0   
4  John Carter is a war-weary, former military ca...   43.926995  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count  \
0                                    Avatar          7.2      11800   
1  Pirates of the Caribbean: At World's End          6.9       4500   
2                                   Spectre          6.3       4466   
3                     The Dark Knight Rises          7.6       9106   
4                               John Carter          6.1       2124   

                                                cast  \
0  Sam Worthington Zoe Saldana Sigourney Weaver S...   
1  Johnny Depp Orlando Bloom Keira Knightley Stel...   
2  Daniel Craig Christoph Waltz L\u00e9a Seydoux ...   
3  Christian Bale Michael Caine Gary Oldman Anne ...   
4  Taylor Kitsch Lynn Collins Samantha Morton Wil...   

                                                crew           director  
0  [{'name': 'Stephen E. Rivkin', 'gender': 0, 'd...      James Cameron  
1  [{'name': 'Dariusz Wolski', 'gender': 2, 'depa...     Gore Verbinski  
2  [{'name': 'Thomas Newman', 'gender': 2, 'depar...         Sam Mendes  
3  [{'name': 'Hans Zimmer', 'gender': 2, 'departm...  Christopher Nolan  
4  [{'name': 'Andrew Stanton', 'gender': 2, 'depa...     Andrew Stanton  

[5 rows x 24 columns]

In [10]:
movie_data.shape

(4803, 24)

In [ ]:
movie_data.describe()

selecting the columns that are used for the recommendation to be done

In [ ]:
selected_features=['genres','keywords','tagline','cast','director']
print(selected_features)

replace the null values with null string. Null string is a empty string(' '). so fillna means null values or not available with empty string.
(fillna) : feature is a variable that
changes like the  iteration and everytime it takes the value as genres,keywords in a loop.
second statement change the feature value which is numm of particular column with null string.
fillna('n/a')=not avaliable will be filled in empty places

In [ ]:
for feature in selected_features:
  movie_data[feature]=movie_data[feature].fillna('')

all features with their data are combined and printed as one value

feature will take the value as genre, keyword, tagline etc and one by one all are filled with null string when value is not available or value is missiing

In [ ]:
movie_data.head()

In [ ]:
combined_features = movie_data['genres']+''+movie_data['keywords']+''+movie_data['tagline']+''+movie_data['cast']+''+movie_data['director']

concatenation is done

In [ ]:
print(combined_features)

now converting the data to the numeric data

In [ ]:
vectorizer=TfidfVectorizer()

In [18]:
feature_vectors=vectorizer.fit_transform(combined_features)

In [19]:
print(feature_vectors)

  (0, 3583)	0.16941894714909375
  (0, 20729)	0.2738578969459924
  (0, 16614)	0.15696073877453268
  (0, 14550)	0.2236809363648068
  (0, 23267)	0.16128139780622516
  (0, 26547)	0.19638671147741732
  (0, 22464)	0.20207089568711745
  (0, 21487)	0.21550128478931546
  (0, 27540)	0.19771357974524176
  (0, 27182)	0.23480088356130552
  (0, 21521)	0.1562356665943368
  (0, 18558)	0.2612168335104853
  (0, 18164)	0.08690831799482265
  (0, 27161)	0.1261988657937151
  (0, 24197)	0.07518543993419265
  (0, 22778)	0.2738578969459924
  (0, 4676)	0.2452909770424961
  (0, 26334)	0.13059723071916296
  (0, 22916)	0.33668756406929184
  (0, 10193)	0.16532432420812987
  (0, 4456)	0.21799675215510664
  (0, 9304)	0.2738578969459924
  (0, 21836)	0.09966592997173944
  (0, 8936)	0.11806131645084653
  (0, 444)	0.09109355212252294
  :	:
  (4801, 5449)	0.31510541707008866
  (4801, 22170)	0.31510541707008866
  (4801, 27361)	0.31510541707008866
  (4801, 27513)	0.3005604008026175
  (4801, 8121)	0.2611505149930138
  (4801,

now we can see text data is convertted from text data to numeric data

now try to get similarity score usine cosine similarity funcction

In [20]:
similarity=cosine_similarity(feature_vectors)

In [21]:
print(similarity)

[[1.         0.06865296 0.01492221 ... 0.         0.         0.        ]
 [0.06865296 1.         0.02799128 ... 0.01243107 0.         0.        ]
 [0.01492221 0.02799128 1.         ... 0.         0.         0.        ]
 ...
 [0.         0.01243107 0.         ... 1.         0.         0.        ]
 [0.         0.         0.         ... 0.         1.         0.        ]
 [0.         0.         0.         ... 0.         0.         1.        ]]


In [22]:
print(similarity.shape)

(4803, 4803)


for one movie similarity score will 4803 and for second movie also 4803 similarity score values.

index and the smilarity score counts


take input from user as movie name

In [23]:
movie_name=input("enter your favourite movie name")

enter your favourite movie nameironman


display a list of movie names

In [24]:
list_of_all_titles=movie_data['title'].tolist()
print(list_of_all_titles)

['Avatar', "Pirates of the Caribbean: At World's End", 'Spectre', 'The Dark Knight Rises', 'John Carter', 'Spider-Man 3', 'Tangled', 'Avengers: Age of Ultron', 'Harry Potter and the Half-Blood Prince', 'Batman v Superman: Dawn of Justice', 'Superman Returns', 'Quantum of Solace', "Pirates of the Caribbean: Dead Man's Chest", 'The Lone Ranger', 'Man of Steel', 'The Chronicles of Narnia: Prince Caspian', 'The Avengers', 'Pirates of the Caribbean: On Stranger Tides', 'Men in Black 3', 'The Hobbit: The Battle of the Five Armies', 'The Amazing Spider-Man', 'Robin Hood', 'The Hobbit: The Desolation of Smaug', 'The Golden Compass', 'King Kong', 'Titanic', 'Captain America: Civil War', 'Battleship', 'Jurassic World', 'Skyfall', 'Spider-Man 2', 'Iron Man 3', 'Alice in Wonderland', 'X-Men: The Last Stand', 'Monsters University', 'Transformers: Revenge of the Fallen', 'Transformers: Age of Extinction', 'Oz: The Great and Powerful', 'The Amazing Spider-Man 2', 'TRON: Legacy', 'Cars 2', 'Green Lant

movie name that we have given will be compared with all movie names and closest match will be found out

now movies name with closest match

In [25]:
find_close_match=difflib.get_close_matches(movie_name,list_of_all_titles)
print(find_close_match)

['Birdman', 'Iron Man', 'Hitman']


on the basis of the name given by the user closest match on the basis of the name will be displayed

among the array of the movie dislayed on the basis of name, first value with index is taken and printed of index 0

In [26]:
close_match=find_close_match[0]
print(close_match)

Birdman


now find the index of the avatar in the list of 4803 movies like it can be 0, 245 or 4000 also in the list of all movies.

In [27]:
index_of_the_movie = movie_data[movie_data.title==close_match]['index'].values[0]
print(index_of_the_movie)

2334


In [28]:
similarity_score=list(enumerate(similarity[index_of_the_movie]))
print(similarity_score)

[(0, 0.01210988245312615), (1, 0.0257067806505408), (2, 0.0), (3, 0.016399444542564753), (4, 0.0), (5, 0.007115903547886342), (6, 0.017582900087314852), (7, 0.05230691401997358), (8, 0.04631644597519841), (9, 0.035236205879916126), (10, 0.0059889951910495004), (11, 0.0), (12, 0.0), (13, 0.033710394105593906), (14, 0.05179892410894137), (15, 0.01654407844079731), (16, 0.05208472819814823), (17, 0.0), (18, 0.0410293326782741), (19, 0.0151694164718962), (20, 0.058712295281856036), (21, 0.0), (22, 0.011589555204584106), (23, 0.008945816146614865), (24, 0.08353692061449355), (25, 0.004220779310035166), (26, 0.0), (27, 0.0), (28, 0.0), (29, 0.0), (30, 0.013862221194848057), (31, 0.010308307168894113), (32, 0.0), (33, 0.0), (34, 0.0), (35, 0.0), (36, 0.0), (37, 0.0375554966447895), (38, 0.055020796194289213), (39, 0.0), (40, 0.016947379128244386), (41, 0.033148589315080544), (42, 0.010679588633355249), (43, 0.005220735651649203), (44, 0.0), (45, 0.0038893326292517767), (46, 0.0199602740430006

above code will print the similarity score with all movies in the liist from 0 index to 4803 indexd movie. that means avatar and pirates of the carribean bichma kati similarity score xa number form ma(80% ...)

means similarity score 4803 movies ko hune bho as compared with avatar

In [29]:
len(similarity_score)

4803

In [30]:
sorted_similar_movies=sorted(similarity_score,key=lambda x:x[1],reverse=True)
print(sorted_similar_movies)

[(2334, 1.0), (2187, 0.2496366204771084), (4089, 0.19699232448189538), (176, 0.1896094298636482), (2133, 0.1757836586310717), (1462, 0.13268090853814288), (2356, 0.13224338559900434), (2086, 0.12640348358377565), (4014, 0.12530714149521943), (2629, 0.11523459518423868), (4580, 0.11318398325318191), (3276, 0.09996924016730382), (615, 0.09971725124051328), (2276, 0.09930618965465274), (283, 0.09444438726817266), (2635, 0.09248467653812097), (1146, 0.08440972857183943), (321, 0.0836427708757282), (24, 0.08353692061449355), (991, 0.08291626484661191), (641, 0.0820550913015031), (2737, 0.08203349286643975), (975, 0.08091339872230989), (427, 0.08044676279375026), (924, 0.0778801458998301), (3951, 0.0777122669564327), (3083, 0.07760477063686813), (1356, 0.07664299697746253), (187, 0.07561693573285624), (2951, 0.07523062260791191), (867, 0.07467515912452041), (3899, 0.07320635390955112), (996, 0.07276825674919836), (1428, 0.07217269649180419), (2627, 0.07203824528502221), (174, 0.0710493675303

x is represented as the similarity score whereas x[1] represents the similarity score value that is 1.00 means . x[0] is the index whereas x[1] is the similarity score. which has to be compared. reverse =true in python make the list in descending order.

0,3158,94 ... are the index of the moviws


movies with similarity score from high to low are displayed. highest score is 1.0

print the name of similar movies using index

movie[0] means it is taking first value from (0,1.000) that means index of the moviw so that title can be extracted from the list.

movie will take one by one value from the sorted similar movies. first value movie will take is (0,1.0..) s0 index =movie[0] means index = 0 index is assigned to index variable from two values. now that index value is used to get the titles.

In [31]:
print('movies suggested for you: \n')
i=1
for movie in sorted_similar_movies:
  index=movie[0]
  title_from_index=movie_data[movie_data.index==index]['title'].values[0]
  if(i<30):
    print(i,'.',title_from_index)
    i+=1

movies suggested for you: 

1 . Birdman
2 . 21 Grams
3 . Amores perros
4 . The Revenant
5 . Babel
6 . Biutiful
7 . Youth in Revolt
8 . Home Alone 2: Lost in New York
9 . Kevin Hart: Let Me Explain
10 . You Will Meet a Tall Dark Stranger
11 . Quinceañera
12 . It's Kind of a Funny Story
13 . Money Train
14 . The Painted Veil
15 . The Taking of Pelham 1 2 3
16 . Funny Games
17 . Man on a Ledge
18 . The Campaign
19 . King Kong
20 . Fair Game
21 . Due Date
22 . St. Vincent
23 . The International
24 . The Hangover Part II
25 . The Ring Two
26 . Tim and Eric's Billion Dollar Movie
27 . While We're Young
28 . The Hangover
29 . Puss in Boots


movie recommendation system

In [32]:
movie_name=input("enter your favourite movie name")

list_of_all_titles=movie_data['title'].tolist()

find_close_match=difflib.get_close_matches(movie_name,list_of_all_titles)

close_match=find_close_match[0]

index_of_the_movie = movie_data[movie_data.title==close_match]['index'].values[0]

similarity_score=list(enumerate(similarity[index_of_the_movie]))

sorted_similar_movies=sorted(similarity_score,key=lambda x:x[1],reverse=True)

print('movies suggested for you: \n')
i=1
for movie in sorted_similar_movies:
  index=movie[0]
  title_from_index=movie_data[movie_data.index==index]['title'].values[0]
  if(i<30):
    print(i,'.',title_from_index)
    i+=1



enter your favourite movie nameiron man
movies suggested for you: 

1 . Iron Man
2 . Iron Man 2
3 . Iron Man 3
4 . The Avengers
5 . Captain America: Civil War
6 . The Helix... Loaded
7 . Avengers: Age of Ultron
8 . X-Men: The Last Stand
9 . X-Men: Days of Future Past
10 . X-Men
11 . X2
12 . Made
13 . Tropic Thunder
14 . The Incredible Hulk
15 . X-Men: Apocalypse
16 . The Amazing Spider-Man 2
17 . Captain America: The Winter Soldier
18 . Gothika
19 . X-Men: First Class
20 . A Scanner Darkly
21 . The Judge
22 . Deadpool
23 . The Best Man
24 . The Nativity Story
25 . Lucky You
26 . The Kite Runner
27 . Arlington Road
28 . Ant-Man
29 . The Hunting Party


In [33]:
from difflib import SequenceMatcher
Str1 = 'pqrs'
Str2 ='pqr'
Seq = SequenceMatcher(a=Str1,b=Str2)
print(Seq.ratio())



0.8571428571428571
